### DISCUSSION TITLE
"_document.js" `ctx` query returns 2 objects

### DISCUSSION THREAD

**Discussion by avaguerrero** [vercel/next.js#15451] [Help] [answered]
Am quite new to Next.js so please help with my understanding. I am in my `_document.js` file and am logging the slug property of `ctx` like so:

```
static async getInitialProps(ctx) {
    const initialProps = await Document.getInitialProps(ctx);

    console.log(ctx.query);
    
    return {
        ...initialProps,
    };
}
```
We have a page called `[slug].js` and I need this slug name to do some API call in `getInitialProps`. However, I get 2 query objects:

```
{ slug: "the-slug-i-need" }
{ slug: 'favicon.ico' }
```

Hence, when I do my api call passing the `slug` parameter, it's taking `favicon.ico` instead of the actual slug. How do I get around this issue? Else, is there any other way to get the URL parameter in "_document.js"?

**Comment by jamesmosier** [comment, marked as answer]
Hi @avaguerrero. For site wide data, using [`_app.js`](https://nextjs.org/docs/advanced-features/custom-app) is recommended. `_document`'s `getInitialProps` is only executed on the server so utilizing `_app` would be better to ensure you get the correct slug as well as allowing it to execute both on the server and client (depending on how you navigated to a page).

**Comment by avaguerrero** [comment]
Hi @jamesmosier thanks for the reply and it answered my question. I just have a follow up. The reason I need to do the api call is to inject a Google Tag Manager code (coming from our server) into the page's `<script>` tag. We're going to have 2 containers. One with fixed GTM code and the other is a custom GTM code. In the `document.js` file, I am able to insert 2 script tags for the GTM containers but I am unable to make the api call (for the custom GTM id) as I mentioned above.

I was able to make this api call if I do it on `_app.js` and I used `TagManager` to insert this script into the DOM. But then the hardcoded script (fixed GTM code) in `_document.js` doesn't work anymore. It seems to be overwritten by whatever is in the `TagManager`.

Are you able to point me into the right direction on how to go about implementing this? Thank you!

**Reply by jamesmosier** [reply]
I'm not certain why it would get overwritten. Is `TagManager` a component of some kind? Or are you referring to just the initialization script for Google Tag Manager? 

Are you using [`next/head`](https://nextjs.org/docs/api-reference/next/head) to inject the scripts dynamically?

This is originally what I was thinking:

```jsx
import Head from 'next/head'

function MyApp({ Component, pageProps }) {
  return (
    <>
      <Head>
        <script type="text/javascript">
          // your fixed GTM code
        </script>

        <script
          dangerouslySetInnerHTML={{
            __html: `${customScriptFromServer}`
          }}
        />
      </Head>
      <Component {...pageProps} />
    </>
  )
}

MyApp.getInitialProps = async (appContext) => {
  const appProps = await App.getInitialProps(appContext);
  const props = {
    ...appProps,
    customScriptFromServer: 'some custom code'
  }

  return { ...props }
}
```

**Reply by avaguerrero** [reply]
Yeah, I realised I can use the `<Head>` component. But your reply confirms that I am going the right direction. Thanks! BTW, GTM also has a corresponding `<noscript>` tag that is currently in `_document.js` `<body>` tag like so:

```
<body>
          <Main />

          <NextScript />
          <noscript>
            <iframe
              src="https://www.googletagmanager.com/ns.html?id=GTM-12345"
              height="0"
              width="0"
              style={{ display: "none", visibility: "hidden" }}
              title="Google Tag Manager"
            ></iframe>
          </noscript>
</body>
```

In the same manner, I need to be able to inject a custom one (besides our fixed one) for the other GTM id. Is there a way to do that insertion via `_app.js`?

**Reply by jamesmosier** [reply]
You could try just putting it under the component:

```jsx
  return (
    <>
      <Head>
        <script type="text/javascript">
          // your fixed GTM code
        </script>

        <script
          dangerouslySetInnerHTML={{
            __html: `${customScriptFromServer}`
          }}
        />
      </Head>
      <Component {...pageProps} />
      
      <noscript>
        <iframe
          src="https://www.googletagmanager.com/ns.html?id=GTM-12345"
          height="0"
          width="0"
          style={{ display: "none", visibility: "hidden" }}
          title="Google Tag Manager"
        ></iframe>
      </noscript>
    </>
  )
}
```
